pyro/infer/mcmc/hmc.py

```diff
@@ -93,6 +109,41 @@ def _get_trace(self, z):
         trace_poutine(*self._args, **self._kwargs)
         return trace_poutine.trace
 
+    @staticmethod
+    def _iter_latent_nodes(trace):
+        for name, node in sorted(trace.iter_stochastic_nodes(), key=lambda x: x[0]):
+            if not (node["fn"].has_enumerate_support or isinstance(node["fn"], _Subsample)):
+                yield (name, node)
+
+    def _compute_trace_log_prob(self, model_trace):
+        if not self._has_enumerable_sites:
+            return model_trace.log_prob_sum()
+        if self.max_iarange_nesting == float("inf"):
+            raise ValueError("Finite value required for `max_iarange_nesting` when model "
+                             "has discrete (enumerable) sites.")
+        model_trace.compute_log_prob()
+        model_trace.compute_score_parts()
+        ordering = {name: frozenset(site["cond_indep_stack"])
+                    for name, site in model_trace.nodes.items()
+                    if site["type"] == "sample"}
+
+        log_probs = defaultdict(float)
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample" and not site["fn"].has_enumerate_support:
```

Comment:
I believe this incorrectly ignores discrete observe sites?

Reply:
Thanks for catching, will change and a test for this.